pkg/fields/selector.go

```diff
@@ -41,144 +44,158 @@ type Selector interface {
 	}
-	return &notHasTerm{field, value}, nil
+	return len(lsel) == 0
 }
 
-func (t *notHasTerm) String() string {
-	return fmt.Sprintf("%v!=%v", t.field, t.value)
+// Add adds requirements to the selector. It copies the current selector returning a new one
+func (lsel internalSelector) Add(reqs ...selectors.Requirement) Selector {
+	var sel internalSelector
+	for ix := range lsel {
+		sel = append(sel, lsel[ix])
+	}
+	for _, r := range reqs {
+		sel = append(sel, r)
+	}
+	sort.Sort(selectors.ByKey(sel))
+	return sel
 }
 
-type andTerm []Selector
-
-func (t andTerm) Matches(ls Fields) bool {
-	for _, q := range t {
-		if !q.Matches(ls) {
+// Matches for a internalSelector returns true if all
+// its Requirements match the input Fields. If any
+// Requirement does not match, false is returned.
+func (lsel internalSelector) Matches(l Fields) bool {
+	for ix := range lsel {
+		if matches := lsel[ix].Matches(l); !matches {
 			return false
 		}
 	}
 	return true
 }
 
-func (t andTerm) Empty() bool {
-	if t == nil {
-		return true
+// String returns a comma-separated string of all
+// the internalSelector Requirements' human-readable strings.
+func (lsel internalSelector) String() string {
+	var reqs []string
+	for ix := range lsel {
+		reqs = append(reqs, lsel[ix].String())
 	}
-	if len([]Selector(t)) == 0 {
-		return true
-	}
-	for i := range t {
-		if !t[i].Empty() {
-			return false
-		}
-	}
-	return true
+	return strings.Join(reqs, ",")
 }
 
-func (t andTerm) RequiresExactMatch(field string) (string, bool) {
-	if t == nil || len([]Selector(t)) == 0 {
-		return "", false
-	}
-	for i := range t {
-		if value, found := t[i].RequiresExactMatch(field); found {
-			return value, found
+// RequiresExactMatch looks over all internal requirements to determine if the field
+// matches on a requirement key that accepts only 1 valid value where operator is IN, ==, =
+func (lsel internalSelector) RequiresExactMatch(field string) (value string, found bool) {
+	for ix := range lsel {
+		if lsel[ix].Key() == field && len(lsel[ix].Values()) == 1 {
+			switch lsel[ix].Operator() {
+			case selectors.DoubleEqualsOperator,
+				selectors.EqualsOperator,
+				selectors.InOperator:
+				return lsel[ix].Values().List()[0], true
+			}
 		}
 	}
 	return "", false
 }
 
-func (t andTerm) Transform(fn TransformFunc) (Selector, error) {
-	next := make([]Selector, len([]Selector(t)))
-	for i, s := range []Selector(t) {
-		n, err := s.Transform(fn)
+func (lsel internalSelector) Transform(fn TransformFunc) (Selector, error) {
+	result := NewSelector()
+	var (
+		err         error
+		newKey      string
+		newValue    string
+		requirement *selectors.Requirement
+	)
+
+	for ix := range lsel {
+		key := lsel[ix].Key()
+		values := lsel[ix].Values()
+
+		newValues := sets.NewString()
+		if len(values) == 0 {
+			newKey, _, err = fn(key, "")
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			for _, value := range values.List() {
+				newKey, newValue, err = fn(key, value)
```

Comment:
What happens if each call returns a different newKey? For example, someone decides to implement a TransformFunc() that recombines keys and values? Let's say key and values were: 

```JSON
'version': ['pod-v1', 'deployment-v1', 'rc-v2']
```

and TransformFunc returns `'v1': 'pod'`, `'v1': 'deployment'`, `'v2': 'rc'`

What happens then?